controller/gip.cpp

```diff
@@ -150,21 +174,46 @@ bool GipDevice::setPowerMode(uint8_t id, PowerMode mode)
     return sendPacket(out);
 }
 
-bool GipDevice::performRumble(RumbleData rumble)
-{
+void GipDevice::queueRumble(const RumbleData rumble) {
+    /* update our back buffer */
+    *tripleBuffer.back = rumble;
+
+    /* make our back buffer the new front buffer */
+    auto frontBuffer = std::atomic_exchange(&tripleBuffer.front, tripleBuffer.back);
+    std::atomic_store(&tripleBuffer.back, frontBuffer);
```

Comment:
Couldn't this be replaced by a single `std::atomic_store(&tripleBuffer.front, tripleBuffer.back)`? If I understood the code correctly, neither the producer nor the consumer *read* from the back buffer. Isn't the back buffer usually used to write the data to directly (making this essentially only double buffering)?

Reply:
No, this is actually a swap operation... These are pointers, we actually write to different memory location each time it passed these commands. The three pointers constantly change to which memory they point to, we are not writing to the same back buffer all the time, we are switching roles of the memory region.

If you just store, you are effectively only have two buffers, and we would need busy waiting before accessing the back buffer again until we know it was released. The third buffer is here so we can work lockless, and do not need to care about atomic updates of a larger buffer. C++ can only swap simple types atomically.

We are having two threads:

The producer thread atomically swaps its back buffer with its front buffer - so now it no longer needs to care about what happens with that buffer. But it can now write its back buffer because nobody else accesses it. If the front buffer has been processed meanwhile is meaningless to the producer: If it filled another buffer, it simply swaps again.

The consumer thread also atomically swaps its back buffer with its front (send) buffer - so it pushes a sent back buffer back to the producer thread.

If you remove one buffer, you need to wait until a buffer was processed: You cannot just swap pointers because that memory may still be accessed by the other thread.

Also, we need those two operations: `std::atomic_exchange` puts the second pointer into the first pointer and returns the old value of the first pointer which is no longer an atomic value. Now, we can store that value safely in our back buffer because we know we are the only one accessing it. This is done with `std::atomic_store`.

You could probably move ownership and make a private buffer in the consumer and producer, so GIP only acts as an intermediate proxy for the third buffer, and only had knowledge of one single buffer then - but you would still need three buffers with proper swapping.

What you are trying to do does not solve this problem: When we just store the back buffer in the front buffer - what buffer are you going to update then? Essentially, the front buffer and the back buffer would now hold the same pointer. This defeats the whole purpose of the consumer/producer architecture: When the consumer accesses this buffer for reading, it may be written to by the other thread at the same time.

The idea behind triple buffering is: We have one buffer we can put in the middle position, which is known to hold a self-consistent, well defined state. Nobody modifies it, nobody owns it. On the left and right side, we have one private buffer each, and each of the threads knows they have exclusive access to their buffer, so they don't need to care about locking which writing or reading contents. If any of those two threads is done with its private buffer, it swaps it atomically with the middle buffer.

All the other locks we need are just there to synchronize between the threads. Also, one thread does it in reverse order to eliminate delaying by one buffer.

Reply:
Ah, your explanation made it much clearer for me. I read a short article about triple buffering before, but I had incorrectly assumed that `std::atomic_exchange` would exchange the actual *value* of that pointer (the `RumbleData`), while it actually exchanges the pointers itself.

> You could probably move ownership and make a private buffer in the consumer and producer [...]

My plan is to extract the triple buffer functionality into a separate class and make it a little bit more generic. Would be great if you could take a quick look at that when I'm finished.

Reply:
`std::atomic_exchange` does not exchange values! It's really only `std::atomic_store`: It copies the second parameter to the first parameter. But it has an additional guarantee of returning the original value of the first parameter atomically. So if anything, it *prepares* a value exchange, it doesn't finalize it.

Reply:
Yes, that's what I meant. I agree that the word *exchange* is a little bit ambiguous, as it doesn't do a swap.

Reply:
> My plan is to extract the triple buffer functionality into a separate class and make it a little bit more generic. Would be great if you could take a quick look at that when I'm finished.

You should first understand the concept and logic of each individual commit to make a generalized abstraction of that otherwise you may accidentally destroy the synchronization between the threads.

It should be quite forward: Simply move the struct into it's own header file, then migrate it to a class with private data, accessor functions and synchronization function at it should work like before. Ownership of the pointers should be managed by this class, that is: It is responsible for allocation and freeing. I used `shared_pointer<>` for this because it will free memory automatically when there's no longer any reference. And only that template is compatible with `atomic`. Unique pointers did not work.

But I was already thinking about it and decided that it's not worth the effort given the current design of the classes, as responsibilities are not fully separated yet. In theory, the input class should not need to know about the buffer and how it is processed: It should just ask for a buffer to fill it, then signal when it's done, maybe by wrapping the writing routine in a lambda.

Reply:
You could then provide a general device buffer class for any buffers handled in the protocol, and make child classes that are implemented as queue, locking single buffer or triple buffer. I don't think you'll need any other buffer type. Then make a method `writeBuffer(...lambda...)` and `readBuffer(...lambda...)`, where lambda is a function that takes a buffer pointer as argument, so your device buffer class does not need to know how to read or write the buffers. It will just provide synchronization and buffer swapping/queueing.

Reply:
> But I was already thinking about it and decided that it's not worth the effort given the current design of the classes, as responsibilities are not fully separated yet. In theory, the input class should not need to know about the buffer and how it is processed: It should just ask for a buffer to fill it, then signal when it's done, maybe by wrapping the writing routine in a lambda. The same for the GIP class but for reading.

I don't think the input class should know about the buffering at all. I believe the controller class should be the one managing the buffer, as that's the one that has to deal with the firmware issues. Other GIP devices *might* not even be affected by this issue (and the GIP class only serves as a basic protocol handler right now).

I've recently discovered that the game input protocol is even more sophisticated than I initially thought: It supports accessories, chunked transfers and querying of RX and TX capabilities (like audio, input type). Implementing those things would require a complete architectural overhaul, which I'm not inclined to do for xow. Those things will land in my new driver.

Reply:
Also keep in mind: Triple buffers are designed to accept loosing or discarding data that cannot be processed in time - which is exactly what we need for the throttled rumble protocol. Other buffer classes would either block on access if the previous buffer isn't done yet, or put it on a queue. Double buffers would lock only when the producer is too fast or the consumer is too slow but would not lock otherwise. So that's not suitable for the rumble protocol.

Reply:
You may look at triple buffering as a three-buffer fifo queue, with one difference: The fifo queue will never discard a buffer, the triple buffer will. In computer graphics, that's the difference between a swap chain and triple buffering: The swap chain will render ahead like triple buffer but it will always present any rendered image, in consequence this means waiting in the worst case. Triple buffers do not have this property: They will just discard buffers that have not been used yet. For the rumble protocol we may discard earlier commands, as a new command will overwrite the controller status completely anyways, and there's no point in feeding it outdated data. And most importantly: You do not want to queue or lock in the thread receiving data from the game, you want to return to it as fast as possible.